pkg/controller/clusterrelocate/clusterrelocate_controller.go

```diff
@@ -0,0 +1,826 @@
+	case *hivev1.DNSZone:
+		t.Status = hivev1.DNSZoneStatus{}
+		if err := replaceOutgoingToIncoming(t); err != nil {
+			return nil, errors.Wrap(err, "could not set relocate status to incoming")
+		}
+	case *hivev1.ClusterDeployment:
+		t.Status = hivev1.ClusterDeploymentStatus{}
+		if err := replaceOutgoingToIncoming(t); err != nil {
+			return nil, errors.Wrap(err, "could not set relocate status to incoming")
+		}
+	default:
+		return nil, errors.Errorf("unknown type to relocate: %T", t)
+	}
+
+	return obj, nil
+}
+
+func clearInstanceSpecificMeta(to metav1.Object) {
+	to.SetSelfLink("")
+	to.SetUID("")
+	to.SetResourceVersion("")
+	to.SetGeneration(0)
+	to.SetCreationTimestamp(metav1.Time{})
+	to.SetOwnerReferences(nil)
+	to.SetClusterName("")
+	to.SetManagedFields(nil)
+}
+
+// replaceOutgoingToIncoming changes the relocate status from outgoing to incoming. This is used when copying a resource
+// to a destination cluster. On the source cluster, the status is outgoing. On the destination cluster, the status is
+// incoming.
+func replaceOutgoingToIncoming(obj hivev1.MetaRuntimeObject) error {
+	relocateName, relocateStatus, err := controllerutils.IsRelocating(obj)
+	if err != nil {
+		return errors.Wrap(err, "could not determine relocate status")
+	}
+	switch relocateStatus {
+	case hivev1.RelocateOutgoing:
+		controllerutils.SetRelocateAnnotation(obj, relocateName, hivev1.RelocateIncoming)
+	case hivev1.RelocateIncoming:
+		// Do nothing as it is already incoming
+	default:
+		return errors.Errorf("resource is not relocating out: status=%q", relocateStatus)
+	}
+	return nil
+}
+
+// ignoreResource determines whether a resource should be ignored during a copy to a destination cluster. An ignored
+// resource will not be copied.
+func (r *ReconcileClusterRelocate) ignoreResource(obj runtime.Object, logger log.FieldLogger) (bool, error) {
+	switch t := obj.(type) {
+	case *corev1.Secret:
+		return r.ignoreSecret(t, logger)
+	default:
+		return false, nil
+	}
+}
+
+// ignoreSecret determines whether a secret should be ignored during a copy to a destination cluster. An ignored
+// secret will not be copied.
+func (r *ReconcileClusterRelocate) ignoreSecret(secret *corev1.Secret, logger log.FieldLogger) (bool, error) {
+	if secret.Type == corev1.SecretTypeServiceAccountToken {
+		logger.Info("ignoring service account token secret")
+		return true, nil
+	}
+	for _, ownerRef := range secret.OwnerReferences {
+		if ownerRef.Kind != "Secret" {
+			continue
+		}
+		owner := &corev1.Secret{}
+		if err := r.Get(context.Background(), client.ObjectKey{Namespace: secret.Namespace, Name: ownerRef.Name}, owner); err != nil {
+			logger.WithField("owner", ownerRef.Name).WithError(err).Log(controllerutils.LogLevel(err), "could not get owner secret")
+			return false, err
+		}
+		if owner.Type == corev1.SecretTypeServiceAccountToken {
+			logger.Info("ignoring secret owned by a service account token secret")
+			return true, nil
+		}
+	}
```

Comment:
In testing, I did something dumb and migrated a cluster in my "hive" namespace. I ended up getting my hiveadmission-serving-cert copied over the one in the other cluster which it believed was a diff. 

I feel like we're wading into waters where hive doesn't really support multiple clusters in one namespace. 

Would it be possible to modify this to pass in the cluster deployment and ignore secrets that are not in one of our known secret fields?

Reply:
That is a TODO, but something that I would like to hold off on.

Reply:
I have not prioritized it since I do not believe that any of our users do multiple clusterdeployments in a single namespace.

Reply:
But even if they did, it would still work. The problem is when you have secrets with the same name in both clusters, which would not happen from just a ClusterDeployment.

Reply:
:+1: